include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -3078,199 +3081,107 @@ namespace internal
-          if (dim == 3)
-            {
-              evaluate_in_face_apply<1, 2>(
-                eval_tangent,
-                eval_normal,
-                values_dofs,
-                fe_eval,
-                scratch_data,
-                evaluation_flag,
-                n_dofs_tangent,
-                std::integral_constant<bool, integrate>());
-            }
-        }
-      else if (face_direction == 1)
-        {
-          // NOTE. Take zx-coordinates into account for dim == 3
-          if (dim == 3)
-            evaluate_in_face_apply<1, 0>(
-              eval_tangent,
-              eval_normal,
-              values_dofs,
-              fe_eval,
-              scratch_data,
-              evaluation_flag,
-              n_dofs_tangent,
-              std::integral_constant<bool, integrate>());
-          else
-            evaluate_in_face_apply<0, 0>(
-              eval_normal,
-              eval_tangent,
-              values_dofs,
-              fe_eval,
-              scratch_data,
-              evaluation_flag,
-              n_dofs_tangent,
-              std::integral_constant<bool, integrate>());
-
-          values_dofs += 3 * n_dofs_tangent;
+      evaluate_in_face_apply<0, EvalNormal, EvalTangent>(
+        values_dofs,
+        fe_eval,
+        scratch_data,
+        evaluation_flag,
+        face_direction,
+        subface_index,
+        std::integral_constant<bool, integrate>());
 
-          evaluate_in_face_apply<-1, 1>(
-            eval_general,
-            eval_general,
-            values_dofs,
-            fe_eval,
-            scratch_data,
-            evaluation_flag,
-            n_dofs_normal,
-            std::integral_constant<bool, integrate>());
-
-          values_dofs += 3 * n_dofs_normal;
-
-          if (dim == 3)
-            {
-              // NOTE. Take zx-coordinates into account
-              evaluate_in_face_apply<0, 2>(
-                eval_normal,
-                eval_tangent,
-                values_dofs,
-                fe_eval,
-                scratch_data,
-                evaluation_flag,
-                n_dofs_tangent,
-                std::integral_constant<bool, integrate>());
-            }
-        }
-      else
-        {
-          evaluate_in_face_apply<0, 0>(
-            eval_normal,
-            eval_tangent,
-            values_dofs,
-            fe_eval,
-            scratch_data,
-            evaluation_flag,
-            n_dofs_tangent,
-            std::integral_constant<bool, integrate>());
-
-          values_dofs += 3 * n_dofs_tangent;
-
-          evaluate_in_face_apply<1, 1>(
-            eval_tangent,
-            eval_normal,
-            values_dofs,
-            fe_eval,
-            scratch_data,
-            evaluation_flag,
-            n_dofs_tangent,
-            std::integral_constant<bool, integrate>());
-
-          values_dofs += 3 * n_dofs_tangent;
+      if (dim == 3)
+        evaluate_in_face_apply<1, EvalTangent, EvalNormal>(
```

Comment:
The order here shouldn't actually matter because I do all the offsets in `evaluate_in_face_apply`. However this makes more sense since I now always put the evaluation of the normal faces last, i.e `face_direction == component` . It also made it easier for me to write everything up on paper beforehand.

Reply:
But what is `face_direction==2` in the 2D case?

Reply:
Ah, the different calls are not with different `face_direction` but rather the `normal_dir` that I pass to the `tensor_evaluation_kernels`. Since I'm using the "general" kernels for the normal faces, this is not used for anything other than the initial offsets of the pointers.

Reply:
Let me reformulate the question: but what is `normal_dir==2` in the 2D case ;)

Reply:
I have to admit that the name might be misleading.. In the `tensor_evaluation_kernels` I use it to indicate in which direction the higher degree is in (which is the same as which component I'm evaluating). In `evaluate_in_face_apply`, other than passing it to the kernels, I also use it to determine the component here https://github.com/dealii/dealii/blob/a3b2bcea0c4dd7f6ef1c33bd8a196a83d72a0f8e/include/deal.II/matrix_free/evaluation_kernels.h#L3162-L3167
but also to shift the pointers here where you can see that I divide by 2 in the 2D case https://github.com/dealii/dealii/blob/a3b2bcea0c4dd7f6ef1c33bd8a196a83d72a0f8e/include/deal.II/matrix_free/evaluation_kernels.h#L3177-L3182
Passing it to the `EvalGeneral` kernels has no effect since the template parameter isn't used in the general tensor evaluator: https://github.com/dealii/dealii/blob/a3b2bcea0c4dd7f6ef1c33bd8a196a83d72a0f8e/include/deal.II/matrix_free/tensor_product_kernels.h#L201-L209
But yes, the name is a bit misleading in this function. I guess I could easily rename it to `component`(?) or something similar. `component` might not be great in this function though.. Let me know if you have any suggestions! The name kinda stuck from when I started thinking of the FE back in February.

Reply:
Although, the name does make sense for anisotropic faces since the normal direction does not correspond to the component, e.g `component = 2`, `face_direction = 1` will have `normal_dir = 0` since we have zx-coordinates. It's just that `normal_dir = 2` does not make sense for a face in 2D as you say, but neither in 3D, but the variable is still useful for the other stuff I mentioned above. Does that help, @peterrum ?

Reply:
(It does not make sense in 3D since the faces are in 2D)

Reply:
What about swapping these?

```cpp
      const unsigned int face_direction = face_no / 2;

      evaluate_in_face_apply<0, EvalGeneral, EvalGeneral>(
        values_dofs,
        fe_eval,
        scratch_data,
        evaluation_flag,
        face_direction,
        subface_index,
        std::integral_constant<bool, integrate>());

      evaluate_in_face_apply<1, EvalNormal, EvalTangent>(
        values_dofs,
        fe_eval,
        scratch_data,
        evaluation_flag,
        face_direction,
        subface_index,
        std::integral_constant<bool, integrate>());

      if (dim == 3)
        evaluate_in_face_apply<2, EvalTangent, EvalNormal>(
          values_dofs,
          fe_eval,
          scratch_data,
          evaluation_flag,
          face_direction,
          subface_index,
          std::integral_constant<bool, integrate>());
```

Reply:
Here

https://github.com/dealii/dealii/blob/0557c001d7c1d28860dba38ee7248aa8c65f1555/include/deal.II/matrix_free/evaluation_kernels.h#L3190-L3198

`normal_dir` can be 2 although the dimension of `EvaluatorTensorProduct` is only 1 or 2...

Reply:
I see you have introduced `normal_dir` in `EvaluatorTensorProduct<evaluate_general>` during the commit https://github.com/NiklasWik/dealii/commit/6918f728e9368e9ff05c6968d60613bdea815074#diff-bd08275179f5b90beff71f4688219f973ca2cb77a58e1df1cb809aff9274fce1. I have overlooked that. Could you revert that change? There is no normal-direction in the general case. I guess the only reason for that was to be able to pass `normal_dir` to the RT type of `EvaluatorTensorProduct`. This does not feel right, since we might need in the future also something else than the normal. I would suggest to encode the information of the normal in the template arguments of the class.

Reply:
> What about swapping these?

I prefer this order since this is the order of the faces in the quadrature points. I think that changing it only will introduce more confusion. Unless it's beneficial from a performance pov, which I doubt..?  I can add a comment explaining the order if you want. 

> Could you revert that change? 

This will require a lot of changes... And I'm not immediately sure how I would make this pretty. I would rather skip this for now, and rather address this if I have time over at the end of my thesis. But if you really think this is necessary now then I'll of course do it!

Reply:
> normal_dir can be 2 although the dimension of EvaluatorTensorProduct is only 1 or 2...

Maybe my explanation above wasn't that good :')